Bug 1706549 - Document that calling Glean.initialize with real values... [doc only] (#1624)
mozilla/glean: docs/user/reference/general/index.md

The following steps are required for applications using the Glean SDK, but not libraries.

{{#include ../../shared/blockquote-info.html}}

##### Note

> The `initialize` function _must_ be called, even if telemetry upload is disabled.
> Glean needs to perform maintenance tasks even when telemetry is disabled, and because Glean
> does this as part of its initialization, it is _required_ to always call the `initialize`
> function. Otherwise, Glean won't be able to clean up collected data, disable queuing of pre-init
> tasks, or perform other required operations.
>
> This does not apply to special builds where telemetry is disabled at build time. In that case, it is acceptable to not call `initialize` at all.

{{#include ../../shared/blockquote-stop.html}}

##### Do not initialize with a fake value for `uploadEnabled`!

> > `Glean.initialize` **must** always be called with real values.

Reviewer: ```suggestion
> `Glean.initialize` must **always** be called with real values.
```

Hmmmm, isn't the `always` the important bit here?
I also think "real" is a bit ... underdefined here. Maybe `correct`?

> > For example, never call `Glean.initialize(upload=true)` if `true` is a placeholder value that later gets reset by `Glean.setUploadEnabled(false)`.

Reviewer: Maybe again lead with the correct example?
Something like:

```suggestion
> Always pass the user preferencen, e.g. `Glean.initialize(upload=userSettings.telemetry_enabled)` or the equivalent for your application.
> Never call `Glean.initialize(upload=true)` if `true` is a placeholder value that later gets reset by `Glean.setUploadEnabled(false)`.
```

The following steps are required for applications using the Glean SDK, but not libraries.

> {{#include ../../shared/blockquote-info.html}}

Author: ```suggestion
{{#include ../../../shared/blockquote-info.html}}
```

##### Note

> The `initialize` function _must_ be called, even if telemetry upload is disabled.
> Glean needs to perform maintenance tasks even when telemetry is disabled, and because Glean
> does this as part of its initialization, it is _required_ to always call the `initialize`
> function. Otherwise, Glean won't be able to clean up collected data, disable queuing of pre-init
> tasks, or perform other required operations.
>
> This does not apply to special builds where telemetry is disabled at build time. In that case, it is acceptable to not call `initialize` at all.

> {{#include ../../shared/blockquote-stop.html}}

Author: ```suggestion
{{#include ../../../shared/blockquote-stop.html}}
```

##### Initialize clean with the correct value for `uploadEnabled`!

> `Glean.initialize` must **always** be called with real values.

> > Always pass the user preferencen, e.g. `Glean.initialize(upload=userSettings.telemetry_enabled)` or the equivalent for your application.

Author: ```suggestion
> Always pass the user preference, e.g. `Glean.initialize(upload=userSettings.telemetry_enabled)` or the equivalent for your application.
```

> Never call `Glean.initialize(upload=true)` if `true` is a placeholder value that later gets reset by `Glean.setUploadEnabled(false)`.

> > Depending on the provided placeholder value, this might trigger the generation of new client ids or the submission of bogus [`deletion-request` pings](../user/pings/deletion_request.md).

Author: ```suggestion
> Depending on the provided placeholder value, this might trigger the generation of new client ids or the submission of bogus [`deletion-request` pings](../../user/pings/deletion_request.md).
```
